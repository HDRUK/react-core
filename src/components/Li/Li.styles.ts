import { css, Theme } from "@emotion/react";
import { getSize } from "../../theme";
import { LiStyleProps } from "./Li.types";

export const root =
    ({ icon }: LiStyleProps) =>
    (theme: Theme) => {
        return css`
            margin-bottom: ${getSize(2, theme)};

            ${icon
                ? `
                display: flex;
                align-items: center;
                gap: ${getSize(1, theme)};
            `
                : ""};
        `;
    };
