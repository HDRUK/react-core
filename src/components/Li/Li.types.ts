import { ComponentProps, ReactNode } from "react";

export interface LiProps extends ComponentProps<"li"> {
    icon?: ReactNode;
}

export type LiStyleProps = Pick<LiProps, "icon">;
