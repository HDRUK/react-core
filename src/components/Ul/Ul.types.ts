import { ComponentProps } from "react";
import { ComponentMarginProps } from "../../types";

export interface UlProps extends ComponentProps<"ul">, ComponentMarginProps {
    listStyleType?: string;
}

export type UlStyleProps = Pick<UlProps, "listStyleType">;
