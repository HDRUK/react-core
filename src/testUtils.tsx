import React from "react";
import { ThemeProvider } from "@emotion/react";
import { render, RenderOptions } from "@testing-library/react";
import { DEFAULT_THEME } from "./theme";

export const Providers: React.FC<{ children: React.ReactNode }> = ({
  children,
}) => {
  return <ThemeProvider theme={DEFAULT_THEME}>{children}</ThemeProvider>;
};

export const renderWithProviders = (
  ui: React.ReactElement,
  options?: RenderOptions
) => {
  return render(ui, { wrapper: Providers, ...options });
};
