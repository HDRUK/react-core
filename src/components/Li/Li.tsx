/** @jsxImportSource @emotion/react */
import { cx } from "@emotion/css";
import * as styles from "./Li.styles";
import { LiProps } from "./Li.types";

const Li = ({ children, className, icon, ...outerProps }: LiProps) => {
    return (
        <li
            className={cx("ui-Li", className)}
            css={[styles.root({ icon })]}
            {...outerProps}>
            {icon} {children}
        </li>
    );
};

export default Li;
