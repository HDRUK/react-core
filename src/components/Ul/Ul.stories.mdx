import { Meta, Story, Canvas } from "@storybook/addon-docs";
import Icon from "../Icon";
import { Card, CardBody } from "../Card";
import Ul from "./Ul";
import Li from "../Li";
import { ReactComponent as CheckIcon } from "../../images/icons/check.svg";
import { ReactComponent as DangerIcon } from "../../images/icons/danger.svg";

<Meta title="Molecules/Ul" component={Ul} />

<Canvas>
    <Story name="Basic">
        {props => (
            <Card>
                <CardBody>
                    <Ul {...props}>
                        <Li>List item 1</Li>
                        <Li>List item 2</Li>
                        <Li>List item 3</Li>
                    </Ul>
                </CardBody>
            </Card>
        )}
    </Story>
    <Story name="With icons">
        {props => (
            <Card>
                <CardBody>
                    <Ul listStyleType="none" {...props}>
                        <Li
                            icon={
                                <Icon
                                    svg={<CheckIcon />}
                                    fill="green600"
                                    size="xl"
                                />
                            }>
                            Icon tick
                        </Li>
                        <Li
                            icon={
                                <Icon
                                    svg={<DangerIcon />}
                                    fill="red600"
                                    size="xl"
                                />
                            }>
                            Icon danger
                        </Li>
                    </Ul>
                </CardBody>
            </Card>
        )}
    </Story>
</Canvas>
