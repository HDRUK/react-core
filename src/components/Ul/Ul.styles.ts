import { css } from "@emotion/react";
import { UlStyleProps } from "./Ul.types";

export const root =
    ({ listStyleType }: UlStyleProps) =>
    () => {
        return css`
            margin: 0;
            padding: 0;
            list-style-position: inside;
            ${listStyleType ? `list-style-type: ${listStyleType};` : ""}
        `;
    };
