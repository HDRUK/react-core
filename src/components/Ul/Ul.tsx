/** @jsxImportSource @emotion/react */
import { cx } from "@emotion/css";
import useCommonStyles from "../../hooks/useCommonStyles";
import * as styles from "./Ul.styles";
import { UlProps } from "./Ul.types";

const Ul = ({
    children,
    mt,
    mb,
    ml,
    mr,
    m,
    mx,
    my,
    className,
    listStyleType,
    ...outerProps
}: UlProps) => {
    const commonStyles = useCommonStyles({
        mt,
        mr,
        ml,
        mb,
        m,
        mx,
        my,
    });

    return (
        <ul
            className={cx("ui-Ul", className)}
            css={[styles.root({ listStyleType }), commonStyles]}
            {...outerProps}>
            {children}
        </ul>
    );
};

export default Ul;
