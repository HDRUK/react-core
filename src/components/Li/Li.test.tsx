import { RenderResult, screen } from "@testing-library/react";
import React from "react";
import { renderWithProviders } from "../../testUtils";
import Ul from "../Ul";
import Li from "./Li";

let wrapper: RenderResult;

describe("Given the Li component", () => {
    describe("When it is rendered", () => {
        beforeEach(() => {
            wrapper = renderWithProviders(
                <Ul>
                    <Li>List item</Li>
                </Ul>
            );
        });

        it("Then should match the snapshot", () => {
            expect(wrapper.asFragment()).toMatchSnapshot();
        });

        it("Then should have the correct content", async () => {
            expect(
                screen.getByRole("listitem", {
                    hidden: true,
                })
            ).toHaveTextContent(/List item/);
        });

        describe("And it has icons", () => {
            beforeEach(() => {
                wrapper.rerender(
                    <Ul>
                        <Li icon="Icon">List item</Li>
                    </Ul>
                );
            });

            it("Then should have the correct content", async () => {
                expect(
                    screen.getByRole("listitem", {
                        hidden: true,
                    })
                ).toHaveTextContent(/Icon List item/);
            });
        });
    });
});
