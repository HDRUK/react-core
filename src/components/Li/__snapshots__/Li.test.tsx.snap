// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Given the Li component When it is rendered Then should match the snapshot 1`] = `
<DocumentFragment>
  <ul
    class="ui-Ul css-rswhjr-Ul"
  >
    <li
      class="ui-Li css-19hw7pz-Li"
    >
       List item
    </li>
  </ul>
</DocumentFragment>
`;
